mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,312 @@
+        self.selection_strategy = selection_strategy
+        self.save_sampled_queries = save_sampled_queries
+        self.aggregate_statistic = MindMeldALClassifier._validate_aggregate_statistic(
+            aggregate_statistic
+        )
+        self.class_level_statistic = MindMeldALClassifier._validate_class_level_statistic(
+            class_level_statistic
+        )
+        self.log_usage_pct = log_usage_pct
+        self.labeled_logs_pattern = labeled_logs_pattern
+        self.unlabeled_logs_path = unlabeled_logs_path
+        self.output_folder = output_folder
+
+        self.results_manager = ResultsManager(output_folder)
+        self.mindmeld_al_classifier = self._get_mindmeld_al_classifier()
+
+        self.init_unsampled_queries_ids = None
+        self.init_sampled_queries_ids = None
+        self.data_bucket = None
+
+    def _get_mindmeld_al_classifier(self):
+        """ Creates an instance of a MindMeld Active Learning Classifier. """
+        return MindMeldALClassifier(
+            self.app_path,
+            self.training_level,
+            self.n_classifiers,
+            self.aggregate_statistic,
+            self.class_level_statistic,
+        )
+
+    @property
+    def num_iterations(self) -> int:
+        """Calculates the number of iterations needed for training.
+        Returns:
+            num_iterations (int): Number of iterations needed for training.
+        """
+        # An additional iteration is added to save training data after the last sampling round.
+        return 1 + math.ceil(len(self.init_unsampled_queries_ids) / self.batch_size)
+
+    @property
+    def __dict__(self):
+        """ Custom dictionary method used to save key experiment params. """
+        return {
+            "app_path": self.app_path,
+            "train_pattern": self.train_pattern,
+            "test_pattern": self.test_pattern,
+            "train_seed_pct": self.train_seed_pct,
+            "n_classifiers": self.n_classifiers,
+            "n_epochs": self.n_epochs,
+            "batch_size": self.batch_size,
+            "training_level": self.training_level,
+            "training_strategies": self.training_strategies,
+            "selection_strategy": self.selection_strategy,
+            "save_sampled_queries": self.save_sampled_queries,
+            "log_usage_pct": self.log_usage_pct,
+            "labeled_logs_pattern": self.labeled_logs_pattern,
+            "unlabled_logs_path": self.unlabeled_logs_path,
+            "output_folder": self.output_folder,
+        }
+
+    def train(self):
+        """Loads the initial data bucket and then trains on every strategy."""
+        logger.info("Creating output folder and saving params.")
+        self.results_manager.create_experiment_folder(
+            active_learning_params=self.__dict__,
+            training_strategies=self.training_strategies,
+        )
+        logger.info("Creating training data bucket.")
```

Comment:
If we are changing "training" to "tuning" (or similar), can we also update this log to use same terminology?

Reply:
Will do

Reply:
Changed